### ISSUE TITLE
False positive in macro expansion (Did not find struct `MyJsClass` parsed before expand #[napi] for impl)

### ISSUE THREAD

**Issue by TheAmniel** [rust-lang/rust-analyzer#17429]
Hi, I'm a bit new to using rust but I like to do a lot of research on what I use and I've had problems with the rust-analyzer as it's a false positive because the code works, disabling macro expansion doesn't mark it as an error.

I was exploring and testing a little bit the possible solutions that I was seeing in the issues of this repository, like this #17231.
But my solution is this [comment](https://github.com/napi-rs/napi-rs/issues/944#issuecomment-997229188_).

~I'm sorry for my bad English~

**rust-analyzer version**: 0.4.1999-standalone (6b8b8ff4c 2024-06-14)
**rustc version**: rustc 1.79.0 (129f3b996 2024-06-10)

**editor or extension**: VSCode Insiders (only rust-analyzer enable)

**relevant settings**:
- CARGO_HOME: `D:\Programs\Rust\.cargo`
- RUSTUP_HOME: `D:\Programs\Rust\.rustup`

**repository link (if public, optional)**: [repo for testing](https://github.com/TheAmniel/analyzer-error)

**code snippet to reproduce**: 
```rust
#[macro_use]
extern crate napi_derive;

#[napi]
pub struct MyJsClass {
  pub input: String,
}

#[napi]
impl MyJsClass {
  // Js constructor
  #[napi(constructor)] // when add this: Did not find struct `MyJsClass` parsed before expand #[napi] for impl 
  pub fn new(input: Option<String>) -> Self {
    MyJsClass {
      input: input.unwrap_or("none".to_string()),
    }
  }

  // class getter
  #[napi(getter)]
  pub fn val(&self) -> &String {
    &self.input
  }

  // class setter
  #[napi(setter)]
  pub fn set_val(&mut self, input: String) {
    self.input = input;
  }

  // class method
  #[napi]
  pub fn get_val(&self) -> String {
    self.input.clone()
  }
}


// But this work
#[napi(constructor)]
pub struct MyJsClass2 {
  // new MyJsClass2("input param")
  pub input: String,
}

#[napi]
impl MyJsClass2 {
  // class getter
  #[napi(getter)]
  pub fn val(&self) -> &String {
    &self.input
  }
  
  // class setter
  #[napi(setter)]
  pub fn set_val(&mut self, input: String) {
    self.input = input;
  }

  // class method
  #[napi]
  pub fn get_val(&self) -> String {
    self.input.clone()
  }
}
```

**Comment by flodiebold** [comment]
`napi` appears to keep state between invocations of the macro and rely on the order the macro is invoked in. This is questionable at best in rustc, and never going to work in rust-analyzer which can lazily expand individual macro calls as needed. The solution in the linked napi issue, setting `rust-analyzer.procMacro.ignored` to ignore the napi macro, is the best you're going to get.

**Comment by Sunny-117** [comment]
"rust-analyzer.procMacro.ignored": { "napi-derive": ["napi"] }
